src/txn/txn.c

```diff
@@ -270,6 +270,72 @@ __wt_txn_bump_snapshot(WT_SESSION_IMPL *session)
     __txn_get_snapshot_int(session, false);
 }
 
+/*
+ * __wt_txn_save_and_update_snapshot --
+ *     Save some of the current transaction's state and update the snapshot.
+ */
+int
+__wt_txn_save_and_update_snapshot(WT_SESSION_IMPL *session, WT_TXN_SAVED_STATE *old_state)
+{
+    WT_TXN *txn;
+    uint64_t *snapshot_buffer;
+
+    txn = session->txn;
+
+    /* Allocate a buffer to save new snapshot array. */
+    WT_RET(__wt_calloc(
+      session, 1, sizeof(txn->snapshot[0]) * S2C(session)->session_size, &snapshot_buffer));
```

Comment:
I do not understand what we are using snapshot_buffer for and why it needs to be "sizeof(txn->snapshot[0]) * S2C(session)->session_size" large. It is likely my ignorance about the snapshot management and transaction code.

Reply:
We want to acquire a new snapshot while reserving the existing snapshot. `snapshot_buffer` is temporary place holder to allocate a new array and assign it `txn::snapshot`. The array should be large enough to hold the number of sessions that can be active (`WT_CONN_IMPL::session_size`).